packages/lib/src/delivery/Messages.ts

```diff
@@ -37,29 +47,82 @@ export async function getMessages(
 
 // buffer message until delivery and sync acknoledgment
 export async function incomingMessage(
-    data: { envelop: EncryptionEnvelop; token: string },
+    { envelop, token }: { envelop: EncryptionEnvelop; token: string },
+    deliveryServiceProfile: DeliveryServiceProfile,
     getSession: (accountAddress: string) => Promise<Session | null>,
     storeNewMessage: (
         conversationId: string,
         envelop: EncryptionEnvelop,
     ) => Promise<void>,
     send: (socketId: string, envelop: EncryptionEnvelop) => void,
 ): Promise<void> {
-    const envelop = {
-        ...data.envelop,
-        deliveryServiceIncommingTimestamp: new Date().getTime(),
-    };
-    const account = formatAddress(formatAddress(data.envelop.from));
-    const contact = formatAddress(formatAddress(data.envelop.to));
-    const conversationId = getConversationId(account, contact);
-
-    if (await checkToken(getSession, account, data.token)) {
-        storeNewMessage(conversationId, envelop);
-        const contactSession = await getSession(contact);
-        if (contactSession?.socketId) {
-            send(contactSession.socketId, envelop);
-        }
-    } else {
+    const sender = formatAddress(envelop.from);
+    const receiver = formatAddress(envelop.to);
+    const conversationId = getConversationId(sender, receiver);
+
+    const tokenIsValid = await checkToken(getSession, sender, token);
+    if (!tokenIsValid) {
+        //Token is invalid
         throw Error('Token check failed');
     }
+
+    const receiverSession = await getSession(receiver);
+    if (receiverSession === null) {
+        //TODO how should we handle this case?
```

Comment:
Is there a possiblity for the session being null ? If so how should we handle such case properly

Reply:
I think we should discuss this on a call

Reply:
Delivery service should only accept users with a session